probatus/missing/imputation.py

```diff
@@ -0,0 +1,318 @@
+                                                          verbose=self.verbose)
+        self.y = preprocess_labels(y, index=self.X.index, verbose=self.verbose)
+                                                         
+
+        #Identify categorical features if not explicitly specified.
+        if 'auto' in categorical_columns:
+            X_cat = X.select_dtypes(include=['category','object'])
+            categorical_columns = X_cat.columns.to_list()
+            for column in categorical_columns:
+                X[column] = X[column].astype('category')
+        else :
+            #Check if the passed columns are in the dataframe.
+            assert categorical_columns in X.columns,"All categorical columns not in the dataframe."
+            X_cat = X[categorical_columns]
+        #Identify the numeric columns.Numeric columns are all columns expect the categorical
+        # columns
+        X_num = X.drop(columns = categorical_columns,inplace=False)
+        numeric_columns = X_num.columns.to_list()
+        
+        for strategy in self.strategies:
+
+            numeric_transformer = Pipeline(steps=[
+                    ('imputer', self.strategies[strategy])])
+
+            categorical_transformer = Pipeline(steps=[
+                    ('imp_cat',SimpleImputer(strategy='constant',fill_value='missing',add_indicator=True)),
+                    ('ohe_cat',OneHotEncoder(handle_unknown='ignore')),
+                ])
+
+            preprocessor = ColumnTransformer(
+                    transformers=[
+                        ('num', numeric_transformer, numeric_columns),
+                        ('cat', categorical_transformer, categorical_columns)],
+                        remainder='passthrough')
+
+            clf = Pipeline(steps=[('preprocessor', preprocessor),
+                                    ('classifier', self.clf)])
+                
+            imputation_results = cross_val_score(
+                clf,
+                X,
+                y,
+                scoring=self.scorer.scorer,
+                cv=self.cv,
+                n_jobs = self.n_jobs)
+
+            temp_results = {
+                    'strategy' : strategy,
+                    'score': np.round(np.mean(imputation_results),3),
+                    'std':np.round(np.std(imputation_results),3),
+                }
+            results.append(temp_results)
+        #If model supports missing values by default, then calculate the scores 
+        #on raw data without any imputation. 
+        if self.model_na_support :
+            categorical_transformer = Pipeline(steps=[
+                    ('ohe_cat',OneHotEncoder(handle_unknown='ignore')),
+                ])
+
+            preprocessor = ColumnTransformer(
+                    transformers=[
+                        ('cat', categorical_transformer, categorical_columns)],
+                        remainder='passthrough')
+
+            self.clf = Pipeline(steps=[('preprocessor', preprocessor),
```

Comment:
Code from here onwards is repeated. Can you merge the two into one?

Reply:
Done.